File: TeeNew/Builder2009/XYZ_Gridding.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "XYZ_Gridding.h"
#include "math.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "Base"
#pragma link "TeePoin3"
#pragma link "TeeSurfa"
#pragma link "TeeTools"
#pragma resource "*.dfm"
TXYZGridding *XYZGridding;
//---------------------------------------------------------------------------
__fastcall TXYZGridding::TXYZGridding(TComponent* Owner)
  : TBaseForm(Owner)
{
}
//---------------------------------------------------------------------------
TChartValue __fastcall TXYZGridding::ClosestValue(TCustom3DSeries *Source, const TChartValue X, const TChartValue Z)
{
  TChartValue res = 0.0;
  double MaxDist = 1.0E+300;
  int Closer = -1;
  TChartValue tmpX, tmpZ, dist;

  for (int t=0; t<Source->Count(); t++)
  {
    tmpX = Source->XValues->Value[t]-X;
    tmpZ = Source->ZValues->Value[t]-Z;
    dist = sqrt(pow(tmpX,2)+pow(tmpZ,2));
    if (dist<MaxDist)
    {
      MaxDist = dist;
      Closer = t;
    }
  }

  if (Closer!=-1) res = Source->YValues->Value[Closer];
  return res;

}
//---------------------------------------------------------------------------
void __fastcall TXYZGridding::Gridding(TCustom3DSeries *Source, TCustom3DGridSeries *Dest, const int GridSize)
{
  TChartValue tmpX, tmpZ;
  TChartValue tmpMinX = Source->XValues->MinValue;
  TChartValue tmpMinZ = Source->ZValues->MinValue;

  TChartValue tmpXFactor =Source->XValues->Range()/GridSize;
  TChartValue tmpZFactor = Source->ZValues->Range()/GridSize;

  Dest->Clear();
  Dest->NumXValues = GridSize;
  Dest->NumZValues = GridSize;

  // loop all grid cells
  for (int x=1; x<Dest->NumXValues; x++)
  {
      tmpX = tmpMinX+((x-1)*tmpXFactor);
      for (int z=1; z<Dest->NumZValues; z++)
      {
        tmpZ = tmpMinZ+((z-1)*tmpZFactor);
        Dest->Value[x][z] = ClosestValue(Source,tmpX,tmpZ); // calculate
      }
  }
}
//---------------------------------------------------------------------------
void __fastcall TXYZGridding::Fill(void)
{
  Series1->Clear();
  int m=50;
  double tmpX, tmpZ;
  for (int x=-m; x<m; x++)
  {
    tmpX = pow(x/30.0,2);
    for (int z=-m; z<m;z++)
    {
      tmpZ = pow(z/30.0,2);
      tmpZ = sqrt(tmpX + tmpZ);
      Series1->AddXYZ(x, 4*cos(3*tmpZ)*exp(-0.5*tmpZ), z);
    }
  }
}
//---------------------------------------------------------------------------

void __fastcall TXYZGridding::FormCreate(TObject *Sender)
{
  Fill();
  Gridding(Series1,Series2,15);
}
//---------------------------------------------------------------------------

void __fastcall TXYZGridding::RadioButton1Click(TObject *Sender)
{
  RadioButton2->Checked = false;
  RadioButton1->Checked = true;

  Series1->Visible = RadioButton1->Checked;
  Series2->Visible = RadioButton2->Checked;

}
//---------------------------------------------------------------------------

void __fastcall TXYZGridding::RadioButton2Click(TObject *Sender)
{
  RadioButton1->Checked = false;
  RadioButton2->Checked = true;

  Series1->Visible = RadioButton1->Checked;
  Series2->Visible = RadioButton2->Checked;
}
//---------------------------------------------------------------------------

File: TeeNew/XYZ_Gridding.pas
unit XYZ_Gridding;
{$I TeeDefs.inc}

interface

uses
  {$IFNDEF LINUX}
  Windows, Messages,
  {$ENDIF}
  SysUtils, Classes,
  {$IFDEF CLX}
  QGraphics, QControls, QForms, QDialogs, QExtCtrls, QStdCtrls, QComCtrls,
  {$ELSE}
  Graphics, Controls, Forms, Dialogs, ExtCtrls, StdCtrls, ComCtrls,
  {$ENDIF}
  Base, TeEngine, TeeSurfa, TeePoin3, TeeProcs, Chart, TeCanvas, TeeTools,
  TeeGDIPlus;

type
  TXYZGridding = class(TBaseForm)
    Label1: TLabel;
    Series1: TPoint3DSeries;
    Series2: TSurfaceSeries;
    RadioButton1: TRadioButton;
    RadioButton2: TRadioButton;
    ChartTool1: TRotateTool;
    procedure FormCreate(Sender: TObject);
    procedure RadioButton1Click(Sender: TObject);
    procedure RadioButton2Click(Sender: TObject);
  private
    { Private declarations }
    Procedure Fill;
  public
    { Public declarations }
  end;

implementation

{$IFNDEF CLX}
{$R *.dfm}
{$ELSE}
{$R *.xfm}
{$ENDIF}

Procedure Gridding(Source:TCustom3DSeries; Dest:TCustom3DGridSeries; GridSize:Integer);
Const MaxDistance=1E+300;

  Function ClosestValue(Source:TCustom3DSeries; const X,Z:TChartValue):TChartValue;
  var Closer,
      t       : Integer;
      tmpX,
      tmpZ,
      dist,
      MaxDist : Double;
  begin
    MaxDist:=MaxDistance;
    Closer:=-1;
    result:=0;

    with Source do
    for t:=0 to Count-1 do
    begin
      tmpX:=XValues.Value[t]-X;
      tmpZ:=ZValues.Value[t]-Z;
      dist:=Sqrt(Sqr(tmpX)+Sqr(tmpZ));
      if dist<MaxDist then
      begin
        MaxDist:=dist;
        Closer:=t;
      end;
    end;

    if Closer<>-1 then
       result:=Source.YValues.Value[Closer];
  end;

var x,
    z : Integer;
    tmpX,
    tmpZ : TChartValue;
    tmpMinX,
    tmpMinZ,
    tmpXFactor,
    tmpZFactor : TChartValue;
begin
  tmpMinX:=Source.XValues.MinValue;
  tmpMinZ:=Source.ZValues.MinValue;

  tmpXFactor:=Source.XValues.Range/GridSize;
  tmpZFactor:=Source.ZValues.Range/GridSize;

  with Dest do
  begin
    Clear;
    NumXValues:=GridSize;
    NumZValues:=GridSize;

    // loop all grid cells
    for x:=1 to NumXValues do
    begin
      tmpX:=tmpMinX+((x-1)*tmpXFactor);

      for z:=1 to NumZValues do
      begin
        tmpZ:=tmpMinZ+((z-1)*tmpZFactor);

        Value[x,z]:=ClosestValue(Source,tmpX,tmpZ); // calculate
      end;
    end;
  end;
end;

// Add many random XYZ points to Series1
Procedure TXYZGridding.Fill;
var tmpX,
    tmpZ : Double;
    m,
    x,
    z : Integer;
begin
  with Series1 do
  begin
    Clear;

    m:=100;

    for x:=-m to m do
    begin
      tmpX:=sqr(x/30);
      for z:=-m to m do
      begin
        tmpZ:=Sqr(z/30);
        tmpZ:=Sqrt(tmpX + tmpZ);
        AddXYZ(x, 4*cos(3*tmpZ)*exp(-0.5*tmpZ), z);
      end;
    end;
  end;
end;

procedure TXYZGridding.FormCreate(Sender: TObject);
begin
  inherited;

  Chart1.View3D := True;
  Fill;

  Gridding(Series1,Series2,15);
end;

procedure TXYZGridding.RadioButton1Click(Sender: TObject);
begin
  RadioButton2.Checked:=False;
  RadioButton1.Checked:=True;

  Series1.Visible:=RadioButton1.Checked;
  Series2.Visible:=RadioButton2.Checked;
end;

procedure TXYZGridding.RadioButton2Click(Sender: TObject);
begin
  RadioButton1.Checked:=False;
  RadioButton2.Checked:=True;

  Series1.Visible:=RadioButton1.Checked;
  Series2.Visible:=RadioButton2.Checked;
end;

initialization
  RegisterClass(TXYZGridding);
end.
